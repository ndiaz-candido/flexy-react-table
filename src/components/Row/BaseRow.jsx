import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  LeftSection,
  RightSection,
  ScrollableSection,
} from './sections';

const defaultRowStyle = {
  display: 'flex',
  overflow: 'hidden',
  minHeight: '50px',
};
const stickyStyle = {
  position: 'sticky',
  top: '0',
  zIndex: '1000',
};
const cellDefaultStyles = width => ({
  width: `${width}px`,
  flex: `0 0 ${width}px`,
});

class BaseRow extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;
  static HEADER_ROW_ID = 0;
  static NOT_SPECIFY_COLUMN_VALUE = 'Not Column';

  state = {
    leftStickySection: [],
    rightStickySection: [],
    scrollableSection: [],
  };

  static getDerivedStateFromProps({ columns, rowData, columnClassName }) {
    const leftStickySection = [];
    const rightStickySection = [];
    const scrollableSection = [];

    columns.forEach(column => {
      const columnWidth = column.width || BaseRow.DEFAULT_COLUMN_WIDTH;
      const columnDataKey = column.dataKey || '';
      const cellContent = rowData[columnDataKey] || '';
      const styles = cellDefaultStyles(columnWidth);
      const cell = (
        <div className={columnClassName} style={styles}>
          {cellContent}
        </div>
      );

      switch (column.stickyAlign) {
        case 'left':
          leftStickySection.push(cell);
          break;
        case 'right':
          rightStickySection.push(cell);
          break;
        default:
          scrollableSection.push(cell);
          break;
      }
    });

    return {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    };
  }

  render() {
    const { className, isSticky, onScroll, rowId, scrollLeft } = this.props;
    const {
      leftStickySection,
      rightStickySection,
      scrollableSection,
    } = this.state;
    const stickyStyleProps = isSticky ? stickyStyle : {};
    const styles = {
      ...defaultRowStyle,
      ...stickyStyleProps,
    };

    return (
      <div className={className} style={styles}>
        <LeftSection cells={leftStickySection} />
        <ScrollableSection
          rowId={rowId}
          cells={scrollableSection}
          onScroll={onScroll}
          scrollLeft={scrollLeft}
        />
        <RightSection cells={rightStickySection} />
      </div>
    );
  }
}

BaseRow.propTypes = {
  columns: PropTypes.array,
  className: PropTypes.any,
  columnClassName: PropTypes.any,
  isSticky: PropTypes.bool,
  isHeader: PropTypes.bool,
  onScroll: PropTypes.func,
  rowId: PropTypes.number,
  rowData: PropTypes.object,
  scrollLeft: PropTypes.number,
};

BaseRow.defaultProps = {
  columns: [],
  className: '',
  columnClassName: '',
  isSticky: false,
  isHeader: false,
  onScroll: () => {},
  rowId: -1,
  rowData: {},
  scrollLeft: 0,
};

export default BaseRow;
