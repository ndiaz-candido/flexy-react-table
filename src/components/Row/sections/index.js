import LeftSection from './LeftSection';
import RightSection from './RightSection';
import ScrollableSection from './ScrollableSection';

export { LeftSection, RightSection, ScrollableSection };
