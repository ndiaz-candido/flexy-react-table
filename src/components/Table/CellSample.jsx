import React, { Component } from 'react';

class CellSample extends Component {
  render() {
    const { collapsed } = this.props;

    return <div> Is collapsed: {`${collapsed}`}</div>;
  }
}

export default CellSample;
