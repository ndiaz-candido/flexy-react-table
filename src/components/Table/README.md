```js
  const TableExample = require('./Table.example.js').default;

  <TableExample />
```
