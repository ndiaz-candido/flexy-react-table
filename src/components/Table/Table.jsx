import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Body from '../Body';
import Row from '../Row';

class Table extends PureComponent {
  tableDefaultStyles = {
    height: 'auto',
    width: '100%',
  };

  getRowsAndColumns = () => {
    const { columns = [], children = [] } = this.props;
    const rows = children.filter(child => child.type === Row);

    return { columns, rows };
  };

  render() {
    const {
      columnHeaderRenderer,
      columnClassName,
      headerRowClassName,
      headerColumnClassName,
      tableClassName,
      rowClassName,
      stickHeader,
    } = this.props;

    return (
      <div className={tableClassName} style={this.tableDefaultStyles}>
        <Body
          columnHeaderRenderer={columnHeaderRenderer}
          columnClassName={columnClassName}
          headerRowClassName={headerRowClassName}
          headerColumnClassName={headerColumnClassName}
          rowClassName={rowClassName}
          stickHeader={stickHeader}
          {...this.getRowsAndColumns()}
        />
      </div>
    );
  }
}

Table.propTypes = {
  /** A set of columns specification */
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,

  /** A function that will be called to render header columns */
  columnHeaderRenderer: PropTypes.func,

  /** The class name to be applied to the table columns */
  columnClassName: PropTypes.any,

  /** The class name to be applied to the header row */
  headerRowClassName: PropTypes.any,

  /** The class name to be applied to each header column */
  headerColumnClassName: PropTypes.any,

  /** The class name to be applied to the table */
  tableClassName: PropTypes.any,

  /** The class name to be applied to the table rows */
  rowClassName: PropTypes.any,

  /** Whether you want a sticky header or not */
  stickHeader: PropTypes.bool,
};

Table.defaultProps = {
  columnHeaderRenderer: null,
  columnClassName: '',
  headerRowClassName: '',
  headerColumnClassName: '',
  tableClassName: '',
  rowClassName: '',
  stickHeader: false,
};

export default Table;
