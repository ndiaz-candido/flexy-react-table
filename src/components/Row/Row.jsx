import React, { Component } from 'react';
import PropTypes from 'prop-types';
import idGenerator from 'react-id-generator';
import { ScrollSection, RightSection, LeftSection } from './sections';
import ScrollContext from '../ScrollContext';

const defaultRowStyle = {
  display: 'flex',
  minHeight: '50px',
};
const defaultCellStyle = width => ({
  width,
  flex: `0 0 ${width}`,
});

class Row extends Component {
  static DEFAULT_COLUMN_WIDTH = 50;

  state = {
    stickRightSection: [],
    stickLeftSection: [],
    scrollSection: [],
  };

  constructor(props) {
    super(props);

    this.rowId = idGenerator();
  }

  static getDerivedStateFromProps({ columns, cellClassName, rowData }) {
    const stickLeftSection = [];
    const stickRightSection = [];
    const scrollSection = [];

    columns.forEach(column => {
      const columnDataKey = column.dataKey;
      const columnWidth = column.width || Row.DEFAULT_COLUMN_WIDTH;
      const cellContent = rowData[columnDataKey];
      const cell = (
        <div className={cellClassName} style={defaultCellStyle(columnWidth)}>
          {cellContent}
        </div>
      );

      switch (column.stickAlign) {
        case 'right':
          stickRightSection.push(cell);
          break;
        case 'left':
          stickLeftSection.push(cell);
          break;
        default:
          scrollSection.push(cell);
          break;
      }
    });

    return {
      stickLeftSection,
      stickRightSection,
      scrollSection,
    };
  }

  render() {
    const { stickLeftSection, stickRightSection, scrollSection } = this.state;
    const { className } = this.props;

    return (
      <div className={className} style={defaultRowStyle}>
        <LeftSection cells={stickLeftSection} />
        <ScrollContext.Consumer>
          {({ handleOnScroll, registerScrollSection }) => (
            <ScrollSection
              cells={scrollSection}
              onScroll={handleOnScroll}
              registerScrollSection={registerScrollSection}
              rowId={this.rowId}
            />
          )}
        </ScrollContext.Consumer>
        <RightSection cells={stickRightSection} />
      </div>
    );
  }
}

Row.propTypes = {
  className: PropTypes.any,
  cellClassName: PropTypes.any,
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      dataKey: PropTypes.string,
      stickAlign: PropTypes.string,
      width: PropTypes.string,
    })
  ),
  rowData: PropTypes.object,
};

Row.defaultProps = {
  className: '',
  cellClassName: '',
  columns: [],
  rowData: {},
};

export default Row;
