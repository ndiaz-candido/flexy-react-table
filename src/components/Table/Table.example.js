/* eslint-disable react/jsx-filename-extension */
import React, { PureComponent } from 'react';
import Table from './Table';
import Row from '../Row';
import CellSample from './CellSample';
import './Table.example.css';

class TableExample extends PureComponent {
  state = {
    isCollapsed: true,
  };

  handleOnCollapsed = () => {
    this.setState({ isCollapsed: !this.state.isCollapsed });
  };

  columnHeaderRenderer = ({ label }) => label;

  render() {
    const { isCollapsed } = this.state;
    const columns = [{
      dataKey: 'a',
      label: 'a',
      stickyAlign: 'left',
      width: '50',
    }, {
      dataKey: 'b',
      label: 'b',
      width: '150',
    }, {
      dataKey: 'c',
      label: 'c',
      width: '150',
    }, {
      dataKey: 'd',
      label: 'd',
      width: '150',
    }, {
      dataKey: 'e',
      label: 'e',
      width: '150',
    }, {
      dataKey: 'f',
      label: 'f',
      width: '150',
    }, {
      dataKey: 'g',
      label: 'g',
      width: '150',
    }, {
      dataKey: 'h',
      label: 'h',
      width: '50',
      stickyAlign: 'right',
    }];

    return (
      <div>
        <div onClick={this.handleOnCollapsed}>Collapse</div>
        <Table
          columns={columns}
          columnHeaderRenderer={this.columnHeaderRenderer}
          columnClassName="column"
          headerRowClassName="headerRow"
          headerColumnClassName="headerColumn"
          tableClassName="table"
          rowClassName="row"
          stickHeader>
          <Row
            rowData={{
              a: <CellSample collapsed={isCollapsed} />,
              b: 'A / B',
              c: 'A / D',
              g: 'A / H',
            }}
          />
          <Row
            rowData={{
              a: 'B / A',
              b: 'B / B',
              c: 'B / C',
            }}
          />
          <Row
            rowData={{
              a: 'C / A',
              b: 'C / B',
              c: 'C / C',
            }}
          />
          <Row
            rowData={{
              a: 'D / A',
              b: 'D / B',
              c: 'D / C',
            }}
          />
        </Table>
      </div>
    );
  }
}

export default TableExample;
