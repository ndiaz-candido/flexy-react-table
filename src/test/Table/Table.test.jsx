import React from 'react';
import { shallow } from 'enzyme';
import Table from '../../components/Table';
import Row from '../../components/Row';

describe('Table', () => {
  test('should render its default content', () => {
    const TableComponent = shallow(<Table columns={[]} />);

    expect(TableComponent).toMatchSnapshot();
  });

  test('should pass down the given columns to the table body', () => {
    const columns = [
      {
        dataKey: 'a',
        label: 'a',
        stickyAlign: 'left',
      },
    ];
    const TableComponent = shallow(<Table columns={columns} />);

    expect(TableComponent).toMatchSnapshot();
  });

  test('should pass down the given rows to the table body', () => {
    const TableComponent = shallow(
      <Table columns={[]}>
        <Row rowData={{}} />
        <Row rowData={{}} />
      </Table>
    );

    expect(TableComponent).toMatchSnapshot();
  });
});
