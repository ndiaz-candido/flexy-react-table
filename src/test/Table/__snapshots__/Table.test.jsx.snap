// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should pass down the given columns to the table body 1`] = `
<div
  className=""
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <Body
    columnClassName=""
    columnHeaderRenderer={null}
    columns={
      Array [
        Object {
          "dataKey": "a",
          "label": "a",
          "stickyAlign": "left",
        },
      ]
    }
    headerColumnClassName=""
    headerRowClassName=""
    rowClassName=""
    rows={Array []}
    stickHeader={false}
  />
</div>
`;

exports[`Table should pass down the given rows to the table body 1`] = `
<div
  className=""
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <Body
    columnClassName=""
    columnHeaderRenderer={null}
    columns={Array []}
    headerColumnClassName=""
    headerRowClassName=""
    rowClassName=""
    rows={
      Array [
        <Row
          rowData={Object {}}
        />,
        <Row
          rowData={Object {}}
        />,
      ]
    }
    stickHeader={false}
  />
</div>
`;

exports[`Table should render its default content 1`] = `
<div
  className=""
  style={
    Object {
      "height": "auto",
      "width": "100%",
    }
  }
>
  <Body
    columnClassName=""
    columnHeaderRenderer={null}
    columns={Array []}
    headerColumnClassName=""
    headerRowClassName=""
    rowClassName=""
    rows={Array []}
    stickHeader={false}
  />
</div>
`;
